Mathlib/Tactic/Bound.lean

```diff
@@ -0,0 +1,280 @@
+them via `have` before calling `bound`.
+
+See `test/Bound.lean` for tests.
+
+### Calc usage
+
+Since `bound` requires the inequality proof to exactly match the structure of the expression, it is
+often useful to iterate between `bound` and `rw / simp` using `calc`.  Here is an example:
+
+```
+-- Calc example: A weak lower bound for `z ↦ z^2 + c`
+lemma le_sqr_add {c z : ℂ} (cz : abs c ≤ abs z) (z3 : 3 ≤ abs z) :
+    2 * abs z ≤ abs (z^2 + c) := by
+  calc abs (z^2 + c)
+    _ ≥ abs (z^2) - abs c := by bound
+    _ ≥ abs (z^2) - abs z := by bound
+    _ ≥ (abs z - 1) * abs z := by rw [mul_comm, mul_sub_one, ← pow_two, ← abs.map_pow]
+    _ ≥ 2 * abs z := by bound
+```
+
+### Aesop rules
+
+`bound` uses threes types of aesop rules: `apply`, `forward`, and closing `tactic`s.  To register a
+lemma as an `apply` rule, tag it with `@[bound]`.  It will be automatically converted into either a
+`norm apply` or `safe apply` rule depending on the number and type of its hypotheses:
+
+1. Nonnegativity/positivity/nonpositivity/negativity hypotheses get score 1 (those involving `0`).
+2. Other inequalities get score 10.
+3. Disjunctions `a ∨ b` gets score 100, plus the score of `a` and `b`.
+
+Score `0` lemmas turn into `norm apply` rules, and score `0 < s` lemmas turn into `safe apply s`
+rules.  The score is roughly lexicographic ordering on the counts of the three type (guessing,
+general, involving-zero), and tries to minimize the complexity of hypotheses we have to prove.
+See `Mathlib.Tactic.Bound.Attribute` for the full algorithm.
+
+To register a lemma as a `forward` rule, tag it with `@[bound_forward]`.  The most important
+builtin forward rule is `le_of_lt`, so that strict inequalities can be used to prove weak
+inequalities.  Another example is `HasFPowerSeriesOnBall.r_pos`, so that `bound` knows that any
+power series present in the context have positive radius of convergence.  Custom `@[bound_forward]`
+rules that similarly expose inequalities inside structures are often useful.
+
+### Guessing apply rules
+
+There are several cases where there are two standard ways to recurse down an inequality, and it is
+not obvious which is correct without more information.  For example, `a ≤ min b c` is registered as
+a `safe apply 4` rule, since we always need to prove `a ≤ b ∧ a ≤ c`.  But if we see `min a b ≤ c`,
+either `a ≤ b` and `a ≤ c` suffices, and we don't know which.
+
+In these cases we declare a new lemma with an `∨` hypotheses that covers the two cases.  Tagging
+it as `@[bound]` will add a +100 penalty to the score, so that it will be used only if necessary.
+Aesop will then try both ways by splitting on the resulting `∨` hypothesis.
+
+Currently the two types of guessing rules are
+1. `min` and `max` rules, for both `≤` and `<`
+2. `pow` and `rpow` monotonicity rules which branch on `1 ≤ a` or `a ≤ 1`.
+
+### Closing tactics
+
+We close numerical goals with `norm_num` and `linarith`.
```

Comment:
I wonder if detecting with the goal is in `Nat` or `Int` and trying `omega` would be helpful. (No need to try this now!)

Reply:
Yes, that's a good idea! I will add a TODO comment.